app/models/ability.rb

```diff
@@ -0,0 +1,31 @@
+# frozen_string_literal: true
+class Ability
+  include Hydra::Ability
+
+  def custom_permissions
+    can :manage, Valkyrie::Model if current_user.admin?
+    alias_action :file_manager, to: :update
+  end
+
+  def read_permissions
+    super
+    can :read, Valkyrie::Model do |obj|
+      valkyrie_test_read(obj) || valkyrie_test_edit(obj)
+    end
+  end
+
+  def edit_permissions
+    super
+    can [:edit, :update, :destroy], Valkyrie::Model do |obj|
+      valkyrie_test_edit(obj)
+    end
+  end
+
+  def valkyrie_test_read(obj)
+    !(user_groups & obj.read_groups).empty? || obj.read_users.include?(current_user.user_key)
```

Comment:
Prefer `.any?` here.

Additionally, consider decomposing into `is_group_readable(obj) || is_user_readable(obj)`. There is context/state binding here that is sort of implied by the ability pattern, but breaking that binding might be helpful (separating the current vs. passed user). Two pairs of explicit/implicit methods (the implicit just passing current_user) might be the most stable long-term.

This could be accomplished with an optional user parameter rather than a pair... A matter of taste, I suppose -- putting the conditional behavior in the method interface.